src/script/cpatch.py

```diff
@@ -0,0 +1,727 @@
+                ],
+            ),
+            (
+                "--python-bindings",
+                "compiled python bindings",
+                [
+                    Component.PYBIND,
+                ],
+            ),
+            (
+                "--python-mgr-modules",
+                "mgr modules written in python",
+                [
+                    Component.PY_MGR,
+                ],
+            ),
+            (
+                "--cephadm",
+                "cephadm command",
+                [
+                    Component.CEPHADM,
+                ],
+            ),
+            (
+                "--dashboard",
+                "dashboard",
+                [
+                    Component.DASHBOARD,
+                ],
+            ),
+            (
+                "--rgw",
+                "radosgw, radosgw-admin",
+                [
+                    Component.RGW,
+                ],
+            ),
+        ]
+        for long_opt, hint, comps in component_selections:
+            AddComponentsAction.add_argument(
+                parser,
+                long_opt,
+                components=comps,
+                hint=hint,
+            )
+
+        return cls(parser.parse_args())
+
+
+class Builder:
+    """Builds an image based on the components added."""
+
+    def __init__(self, cli_ctx):
+        self._ctx = cli_ctx
+        self._jobs = []
+        self._include_dashboard = False
+        self._cached_py_site_packages = None
+        self._workdir = pathlib.Path(
+            tempfile.mkdtemp(suffix=".cpatch", dir=self._ctx.build_dir)
+        )
+        log.debug("cpatch working dir: %s", self._workdir)
+
+    def add(self, component):
+        """Request that the given component be added to the build."""
+        log.info("Including: %s", component)
+        dispatch = {
+            Component.PY_COMMON: self._py_common_job,
+            Component.CEPHADM: self._cephadm_job,
+            Component.PY_MGR: self._py_mgr_job,
+            Component.DASHBOARD: self._py_mgr_job,
+            Component.PYBIND: self._pybind_job,
+            Component.CORE: self._core_job,
+            Component.CEPHFS: self._cephfs_job,
+            Component.RBD: self._rbd_job,
+            Component.RGW: self._rgw_job,
+        }
+        job = dispatch[component]
+        if component == Component.DASHBOARD:
+            self._include_dashboard = True
+        if job not in {j for _, j in self._jobs}:
+            self._jobs.append((component, job))
+
+    def build(self):
+        """Build the container image."""
+        dlines = [f"FROM {self._ctx.base_image}"]
+        jcount = len(self._jobs)
+        for idx, (component, job) in enumerate(self._jobs):
+            num = idx + 1
+            log.info(f"Executing job {num}/{jcount} for {component}")
+            dresult = job(component)
+            dlines.extend(dresult)
+
+        with open(self._workdir / "Dockerfile", "w") as fout:
```

Comment:
Does this mean I should make sure not to have a Dockerfile in my working dir (which I think is defaulting to `ceph/src` here)? Just asking since I currently do have a Dockerfile I keep in `ceph/src` I use for making images with python changes copied in.

Reply:
workdir != <your CWD>.  This is cpatch's working directory, and that's a temporary directory.